server/hotrod/src/main/java/org/infinispan/server/hotrod/ClientListenerRegistry.java

```diff
@@ -320,20 +320,36 @@ void init() {
                // Remove the listener, but do it on another thread pool to not exhaust the IO thread pool
                addListenerExecutor.submit(() -> {
                   log.debug("Channel disconnected, removing event sender listener for id: " + Util.printArray(listenerId));
-                  cache.removeListener(this);
+                  cache.removeListenerAsync(this);
```

Comment:
My understanding of this is that `removeListenerAsync` will execute on the `ClusterExecutor` .. So does it make sense for this still to be called form the `addListenerExecutor` now that it's async?

Could we not do something like:

```java
ch.closeFuture().addListener(f -> {
            log.debug("Channel disconnected, removing event sender listener for id: " + Util.printArray(listenerId));
            cache.removeListenerAsync(this).thenRun(this::unblockCommands);
         });
```
cache.removeListenerAsync(this);

Reply:
No, I agree. Good catch. Let me change this.

Reply:
Also just to clarify you cannot use `thenRun` in this case. We need to ensure we always invoke `unblockCommands` so it would need to use `whenComplete`.

Reply:
Of course, I was forgetting that `thenRun` only executes on successful completion.